File: medium/152. Maximum Product Subarray/maxProduct.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
    result = Math.max(...nums)
    localMax = 1
    localMin = 1

    nums.forEach(x => {
        if (x == 0) {
            localMax = 1
            localMin = 1
        }

        tmpMax = localMax
        localMax = Math.max(x * localMax, x * localMin, x)
        localMin = Math.min(x * tmpMax, x * localMin, x)

        result = Math.max(result, localMax)
    })
    return result
};

console.log(maxProduct([2, 3, -2, 4]))

File: medium/152. Maximum Product Subarray/max_product.py
def max_product(nums):
    result = max(nums)
    local_max = 1
    local_min = 1

    for num in nums:
        if num == 0:
            local_max = 1
            local_min = 1
            continue

        tmp_max = local_max

        local_max = max(num * local_max, num * local_min, num)
        local_min = min(num * tmp_max, num * local_min, num)

        result = max(local_max, result)

    return result


print(max_product([2, 3, -2, 4]))
